sre-recipes/recipe_runner.py

```diff
@@ -0,0 +1,196 @@
+            raise ValueError("Cannot parse config as YAML.")
+
+    def get_name(self):
+        return self.recipe.get("name", "No name found")
+
+    def get_description(self):
+        return self.recipe.get("description", "No description found")
+
+    ############################ Run Recipe ###################################
+
+    @ property
+    def config(self):
+        return self.recipe.get("config", {})
+
+    def run_break(self):
+        print('Deploying broken service...')
+        self.__handle_actions(self.config.get("break", {}))
+        print('Done. Deployed broken service')
+
+    def run_restore(self):
+        print('Restoring broken service...')
+        self.__handle_actions(self.config.get("restore", {}))
+        print('Done. Restored broken service to working state.')
+
+    def run_hint(self):
+        hint = self.config.get("hint", None)
+        if hint:
+            print(f'Here is your hint!\n\n{hint}')
+        else:
+            print("This recipe has no hints.")
+
+    def run_verify(self):
+        verify_config = self.config.get("verify", {})
+        if not verify_config:
+            raise NotImplementedError("Verify is not configured")
+
+        affected_service_config = verify_config.get("affected_service", {})
+        if affected_service_config:
+            if "answer" not in affected_service_config:
+                raise ValueError(
+                    "Correct answer is not specified for affected service quiz.")
+            elif "choices" not in affected_service_config:
+                raise ValueError(
+                    "No answer choices configured in affected service quiz.")
+            utils.run_interactive_multiple_choice(
+                "Which service has an issue?",
+                affected_service_config["choices"],
+                affected_service_config["answer"])
+
+        incident_cause_config = verify_config.get("incident_cause", {})
+        if incident_cause_config:
+            if "answer" not in incident_cause_config:
+                raise ValueError(
+                    "Correct answer is not specified for incident cause quiz.")
+            elif "choices" not in incident_cause_config:
+                raise ValueError(
+                    "No answer choices configured in incident cause quiz.")
+            utils.run_interactive_multiple_choice(
+                "What was the cause of the issue?",
+                incident_cause_config["choices"],
+                incident_cause_config["answer"])
+
+    ########################## Recipe Action Handlers ##########################
+
+    def __handle_actions(self, actions):
+        """
+        Dispatch and handle a list of actions synchronously.
+
+        Paramters
+        ---------
+        actions: a list of dictionary of paramters.
+            Example: [{'run': 'echo "Hello World!"'}]
+        """
+        loadgen_ip = None
+
+        for action in actions:
+            if "run" in action:
```

Comment:
Should the actions have a `type` field? It seems like there are two types (`run` and `loadgen`), but the type is inferred by searching for a specific key in the dictionary, which might cause issues if we add more options over time.

Reply:
Building off that suggestion:
- instead of `type`, `action` might also work well in YAML as the field name
- `loadgen_stop` and `loadgen_spawn` could be separate types, since `stop` doesn't need the arguments that `start` does
- `run` could take a dequence of commands to cut out some extra steps

Example
```
  break:
    - action: loadgen-stop
    - action: run
      commands: 
        - kubectl set env deployment/frontend CONVERT_CURRENCIES=true
        - kubectl delete pod $(kubectl get pod -l app=frontend -o jsonpath='{.items[0].metadata.name}')
        - kubectl wait --for=condition=available --timeout=600s deployment/frontend
    - action: loagen-spawn
      user_type: BasicHomePageViewingUser
      user_count: 20
      spawn_rate: 5
      stop_after: 600
```

Reply:
Done.